src/C-interface/dense/bml_diagonalize_dense.c

```diff
@@ -85,6 +92,135 @@ bml_diagonalize_dense_magma_single_real(
+    float *A_matrix = (float *) A->matrix;
+    float *typed_eigenvalues = (float *) eigenvalues;
+
+    float *evecs = (float *) malloc(A->N * A->N * sizeof(float));
+    float *evals = (float *) malloc(A->N * sizeof(float));
+
+#pragma omp target enter data map(alloc:evecs[0:N*N])
+#pragma omp target enter data map(alloc:evals[0:N])
+// pull from GPU
+#pragma omp target update from(A_matrix[0:N*N])
+    printf("Checking A matrix values \n");
+    for (int i = 0; i < N; i++)
+    {
+        for (int j = 0; j < N; j++)
+        {
+            printf(" %6.3f", A_matrix[i * N + j]);
+        }
+        printf("\n");
+    }
+    // copy A to evecs on GPU
+#pragma omp target teams distribute parallel for
+    for (int i = 0; i < N*N; i++)
+    {
+        evecs[i] = A_matrix[i];
+    }
+#pragma omp target teams distribute parallel for
+    for (int i = 0; i < N; i++)
+    {
+        evals[i] = 1.0;
+    }
+
+#pragma omp target update from(evecs[0:N*N])
+#pragma omp target update from(evals[0:N])
+    // Check values before syev
+    printf("Checking evecs input values \n");
+    for (int i = 0; i < N; i++)
+    {
+        for (int j = 0; j < N; j++)
+        {
+            printf(" %6.3f,", evecs[i * N + j]);
+        }
+        printf("\n");
+    }
+    for (int i = 0; i < N; i++)
+    {
+        printf("%d, %f \n", i, evals[i]);
+    }
+#ifdef BML_SYEVD
+    // Divide and conquer solver
+    MKL_INT lwork = 1 + 6 * N + 2 * N * N;
+    float *work = (float *) malloc(lwork * sizeof(float));
+    MKL_INT liwork = 3 + 5 * N;
+    MKL_INT *iwork = (MKL_INT *) malloc(liwork * sizeof(MKL_INT));
+#pragma omp target enter data map(alloc:work[0:lwork])
+#pragma omp target enter data map(alloc:iwork[0:liwork])
+
+#pragma omp target variant dispatch use_device_ptr(evecs, evals, work, iwork)
+    ssyevd("V", "U", &N, evecs, &N, evals, work, &lwork, iwork, &liwork, &info);
+    free(iwork);
+#else
+    MKL_INT lwork = 3 * N;
+    float *work = (float *) mkl_malloc(lwork * sizeof(float), 64);
+#pragma omp target enter data map(alloc:work[0:lwork])
+#pragma omp target variant dispatch device(dnum) use_device_ptr(evecs, evals, work)
+    ssyev("V", "U", &N, evecs, &N, evals, work, &lwork, &info);
+#endif // BML_SYEVD
+
+#ifdef MKL_GPU_DEBUG
+#pragma omp target update from(evecs[0:N*N])
+#pragma omp target update from(evals[0:N])
+
+    printf("After SYEV \n");
+    for (int i = 0; i < N; i++)
+    {
+        for (int j = 0; j < N; j++)
+        {
+            printf(" %6.3f", evecs[i * N + j]);
+        }
+        printf("\n");
+    }
+    for (int i = 0; i < N; i++)
+    {
+        printf("%d, %f \n", i, evals[i]);
+    }
+#endif
+
+// need typed_eigenvalues on CPU
+#pragma omp target update from(evals[0:N])
+    for (int i = 0; i < N; i++)
+    {
+        typed_eigenvalues[i] = (float) evals[i];
+    }
+
+// leave eigenvectors on GPU
+    float *e_matrix = (float *) eigenvectors->matrix;
+#pragma omp target enter data map(alloc:e_matrix[0:N*N])
```

Comment:
Isn't this array already allocated? We just need to copy transposed results into it...

Reply:
There is no gpu array for eigenvectors allocated yet, unless I missed it. This was the original confusing reuse of A_matrix in the original code

Reply:
Two arrays are passed in as arguments: A and eigenvectors. They are both BML matrices, and thus should have been allocated both on cpu and gpu when created. A contains the matrix (input), eigenvectors contains the computed eigenvectors (output). Since *syev* takes only one matrix argument as an input and output, we use "eigenvectors" for that, and copy A into eigenvectors before calling *syev*

Reply:
Why isn't this done in the original code (evecs is allocated and used instead of eigenvectors)? I copied the original code as much as possible

Reply:
It is true that in the magma code I introduced a third matrix- that maybe could be avoided. But the main issue here is that in the MKL_GPU code, eigenvectors should not be reallocated, but should be assigned to.

Reply:
I think the only thing to change here is simply to remove the allocation for e_matrix on GPU. And I think the third matrix is indeed necessary to do the transpose without touching the originl A.